x/valset-pref/validator_set.go

```diff
@@ -115,6 +121,88 @@ func (k Keeper) UndelegateFromValidatorSet(ctx sdk.Context, delegatorAddr string
+// 2. the (re)delegator is attempting to create a new redelegation where the source validator for this new redelegation is Validator X
+func (k Keeper) PreformRedelegation(ctx sdk.Context, delegator sdk.AccAddress, existingSet types.ValidatorSetPreferences, newSet []types.ValidatorPreference) error {
+	var existingValSet []valSet
+	var newValSet []valSet
+	totalTokenAmount := sdk.NewDec(0)
+
+	// Rearranging the exisingValSet and newValSet to to add extra validator padding
+	for _, existingVals := range existingSet.Preferences {
+		valAddr, validator, err := k.GetValidatorInfo(ctx, existingVals.ValOperAddress)
+		if err != nil {
+			return err
+		}
+
+		// check if the user has delegated tokens to the valset
+		delegation, found := k.stakingKeeper.GetDelegation(ctx, delegator, valAddr)
+		if !found {
+			return fmt.Errorf("No delegation found")
+		}
+
+		tokenFromShares := validator.TokensFromShares(delegation.Shares)
+		existing_val, existing_val_zero_amount := k.GetValSetStruct(existingVals, tokenFromShares)
+		existingValSet = append(existingValSet, existing_val)
+		newValSet = append(newValSet, existing_val_zero_amount)
+		totalTokenAmount = totalTokenAmount.Add(tokenFromShares)
+	}
+
+	for _, newVals := range newSet {
+		amountToDelegate := newVals.Weight.Mul(totalTokenAmount)
+
+		new_val, new_val_zero_amount := k.GetValSetStruct(newVals, amountToDelegate)
+		newValSet = append(newValSet, new_val)
+		existingValSet = append(existingValSet, new_val_zero_amount)
+	}
+
+	// calculate the difference between two sets
+	var diffValSet []*valSet
+	for i, newVals := range existingValSet {
+		diffAmount := newVals.amount.Sub(newValSet[i].amount)
+
+		diff_val := valSet{
+			valAddr: newVals.valAddr,
+			weight:  newVals.amount,
+			amount:  diffAmount,
+		}
+		diffValSet = append(diffValSet, &diff_val)
+	}
+
+	// Algorithm starts here
+	for _, diff_val := range diffValSet {
+		for diff_val.amount.GT(sdk.NewDec(0)) {
+			source_large := diff_val.valAddr
+			target_large, idx := k.FindMin(diffValSet)
+
+			validator_source, err := sdk.ValAddressFromBech32(source_large)
+			if err != nil {
+				return fmt.Errorf("validator address not formatted")
+			}
+
+			validator_target, err := sdk.ValAddressFromBech32(target_large.valAddr)
+			if err != nil {
+				return fmt.Errorf("validator address not formatted")
+			}
+
+			amount := sdk.MinDec(target_large.amount.Abs(), diff_val.amount)
+			_, err = k.stakingKeeper.BeginRedelegation(ctx, delegator, validator_source, validator_target, amount)
+			if err != nil {
+				return err
+			}
+
+			// Find target value in diffValSet and set that to (sourceAmt - targetAmt)
+			diff_val.amount = diff_val.amount.Sub(amount)            // set the source to 0
+			diffValSet[idx].amount = target_large.amount.Add(amount) // set the target to (sourceAmt - targetAmt)
+		}
+	}
+
+	return nil
+}
```

Comment:
I think this code block deserves some in line comments

Reply:
just added